utils/apollo/schema/resolvers.js

```diff
@@ -1,54 +1,130 @@
+      try {
+        users = await User.find(filter);
+        console.debug('users', users);
+      } catch (err) {
+        console.debug(err);
       }
-      return data;
+
+      return users;
     },
 
     spaces: async (parent, args, context) => {
       const { spaceIds } = args;
       console.debug('args', args, 'spaceIds', spaceIds);
-      const data = [];
 
-      await dbConnect();
+      // Fetch all spaces
+      if (spaceIds?.length == 0) {
+        const spaces = await Space.find({}).populate('participants').exec();
+        return spaces;
+      }
 
-      for (let spaceId of spaceIds) {
-        try {
-          const space = await Space.findOne({ spaceId });
-          data.push(space);
-        } catch (err) {
-          console.debug(err);
-        }
+      // For Query by ID
+      const filter = {
+        spaceId: {
+          $in: spaceIds,
+        },
+      };
+
+      let spaces = {};
+      try {
+        spaces = await Space.find(filter).populate('participants').exec();
+      } catch (err) {
+        console.debug(err);
+      }
+      return spaces;
+    },
+  },
+  Mutation: {
+    createSpace: async (parent, args, context) => {
+      console.debug('createSpace', args);
+      const { input } = args;
+      const { name, description, userId, spaceId } = input;
+
+      let result = {};
+      try {
+        const space = new Space({
+          name,
+          description,
+          participants: [],
+          spaceId,
+          isActive: true,
+          music: '',
+        });
+
+        result = await space.save();
+        console.debug('new space', result);
+      } catch (err) {
+        console.debug('Cannot upload new space to database: ', err);
+      }
+      return result;
+    },
+    addUserToSpace: async (parent, args, context) => {
+      console.debug('AddUserToSpace', args);
+      const {
+        input: { spaceId, userId },
+      } = args;
+
+      let space = {};
+      try {
+        // The new option returns the space after the update
+        const filter = { spaceId };
+        const update = {
+          $push: { participants: userId },
+        };
+        space = await Space.findOneAndUpdate(filter, update, { new: true }).populate('participants').exec();
+        console.debug('Updated space after populating participant field:', space);
+      } catch (err) {
+        console.debug('Cannot upload new space to database: ', err);
+      }
+      return space;
+    },
+    removeUserFromSpace: async (parent, args, context) => {
+      console.debug('removeUserFromSpace', args);
+      const {
+        input: { spaceId, userId },
+      } = args;
+
+      let space = {};
+      try {
+        // The new option returns the space after the update
+        const filter = { spaceId };
+        const update = {
+          $pull: { participants: userId },
+        };
+        space = await Space.findOneAndUpdate(filter, update, { new: true }).populate('participants').exec();
```

Comment:
stupid question but I actually don't know. What does `$pull` do? Can you send documentation my way :3 Also, why do you have the new: true option here too lol.

Reply:
$pull and $push are operators that deal with array elements. MongoDB returns the object before it was updated by default, so setting new:true returns the object after it was updated instead. 
https://docs.mongodb.com/manual/reference/operator/update/pull/

Reply:
lovely. Thanks for the update :D